core/app/models/spree/refund.rb

```diff
@@ -8,14 +8,11 @@ class Refund < Spree::Base
 
     validates :payment, presence: true
     validates :reason, presence: true
-    validates :transaction_id, presence: true, on: :update # can't require this on create because the before_create needs to run first
+    validates :transaction_id, presence: true, on: :update # can't require this on create, #perform! provides the transaction_id
```

Comment:
We should either remove the `on:` condition here or have a more robust way of determining whether a refund needs to have a transaction id.  With the updates from this PR we no longer need the `on:` for Solidus usage since the transaction id is fetched before the record is validated.

I'd be inclined to just remove the `on:` condition and require that things like data imports supply a manufactured id of some sort, if they don't have one already.

Reply:
I think opting to remove this validation might be the best route. It allows us to `create` refunds before calling `perform!` on them, rather than building them without saving to the database. It also allows data imports of refunds without `transaction_id`s, which is a potentital scenario. 

The error handling within the `#perform!` call should be enough to alert users that a refund transaction was unsuccessful, as a `transaction_id` is required for the method to succeed.

Reply:
Thank you for all your input on these changes Jordan, you've helped shed a lot of light on potential vulnerabilities with a change such as this. I know adding state to the refund model is something that not everyone will agree with.
Ultimately for the average user these changes won't have any negative implications, but for some users they will have more power over how their store handles refunds.